io/io/src/TBufferJSON.cxx

```diff
@@ -1952,13 +1953,16 @@ void TBufferJSON::JsonReadTObjectMembers(TObject *tobj, void *node)
    UInt_t bits = json->at("fBits").get<unsigned>();
    // UInt32_t pid = json->at("fPID").get<unsigned>(); // ignore PID for the moment
 
+   static auto tobj_fbits_offset = TObject::Class()->GetDataMemberOffset("fBits");
+
    tobj->SetUniqueID(uid);
-   // there is no method to set all bits directly - do it one by one
-   for (unsigned n = 0; n < 32; n++)
-      tobj->SetBit(BIT(n), (bits & BIT(n)) != 0);
 
-   if (gDebug > 2)
-      Info("JsonReadTObjectMembers", "Reading TObject part bits %u kMustCleanup %d", bits, tobj->TestBit(kMustCleanup));
+   // there is no method to set all bits directly - do it differently
+   if (tobj_fbits_offset > 0) {
+      UInt_t *fbits = (UInt_t *) ((char* ) tobj + tobj_fbits_offset);
+      *fbits = bits | TObject::kIsOnHeap | TObject::kNotDeleted;
```

Comment:
Isn't that marking all read object has beeing on the heap?
Wouldn't 
```
UInt_t *fbits = (UInt_t *) ((char* ) tobj + tobj_fbits_offset);
bool isonheap = fbits & TObject::kIsOnHeap;
*fbits = bits | TObject::kNotDeleted;
if (isonheap)
   *fbits = *fbits | TObject::kIsOnHeap ;
else
   *fbits = *fbits & ~TObject::kIsOnHeap;
```
be 'better'?

Reply:
I did it like in `TObject::Streamer()` method. `kIsOnHeap` always set when object read from TBuffer.

Reply:
That is a fair point ;).  I think that since this code was written, the comment:
```
      fBits |= kIsOnHeap;  // by definition de-serialized object is on heap
```
is no longer strictly true (and the same problem is likely in TStreamerInfoReadBuffer.cxx and the StreamerInfoActions) as either via the 'newer' interface in TDirectory where you can pass an existing object or via TBranch's SetAddress one could pass an object on the stack to be read into ....

Reply:
Ok, I will preserve kIsOnHeap bit assigned before in TObject constructor.

Reply:
Done in https://github.com/root-project/root/pull/4182/commits/7269e1e15b64b2240f4f8101189679386b02b1f7